sympy/vector/functions.py

```diff
@@ -196,24 +216,68 @@ def gradient(scalar, coord_sys):
     scalar : SymPy Expr
         The scalar field to compute the gradient of
 
-    coord_sys : CoordSysCartesian
-        The coordinate system to calculate the gradient in
-
     Examples
     ========
 
-    >>> from sympy.vector import CoordSysCartesian, gradient
-    >>> R = CoordSysCartesian('R')
+    >>> from sympy.vector import CoordSystem3D, gradient
+    >>> R = CoordSystem3D('R')
     >>> s1 = R.x*R.y*R.z
-    >>> gradient(s1, R)
+    >>> gradient(s1)
     R.y*R.z*R.i + R.x*R.z*R.j + R.x*R.y*R.k
     >>> s2 = 5*R.x**2*R.z
-    >>> gradient(s2, R)
+    >>> gradient(s2)
     10*R.x*R.z*R.i + 5*R.x**2*R.k
 
     """
+    # Check input
+    if isinstance(scalar, Zero) or scalar == 0:
+        return Vector.zero
+
+    if not issubclass(type(scalar), Expr) or scalar.is_Vector:
+        raise TypeError("Your input should be a scalar field.")
+
+    for atom in scalar.atoms():
+        if isinstance(atom, BaseScalar):
+            system = atom.system
+            break
```

Comment:
> There is a function which derives the metric from the transformation.

Can you point me to it?  I can't seem to find it.

> And... no, deriving the differential operators from the sets of transformations requires usage of equation solvers beyond the current capability of SymPy.

Don't you feel that we should implement this before we continue?  For the purposes of scalability, I honestly feel this is the best route; not to mention the basic building blocks for all coordinate system transformations.  If we have a good curvilinear coordinate class the rest is cake, and I really don't see how speed is going to be an issue, at least in dim 3.  It's only a few partial derivatives and a square root to get the Lamé coefficients.